fedimg/consumers.py

```diff
@@ -93,13 +101,60 @@ def _get_upload_urls(self, builds):
         return upload_files
 
     def consume(self, msg):
-        """ This is called when we receive a message matching the topic. """
+        """ This is called when we receive a message matching our topics. """
 
-        builds = list()  # These will be the Koji build IDs to upload, if any.
+        log.info('Received %r %r' % (msg['topic'], msg['body']['msg_id']))
 
-        msg_info = msg["body"]["msg"]["info"]
+        if msg['topic'].endswith('.task.state.change'):
+            # Scratch tasks.. the old way.
+            return self._consume_scratch_task(msg)
+        elif msg['topic'].endswith('.build.state.change'):
+            # Full builds from pungi4.. the new way.
+            return self._consume_full_build(msg)
+        else:
+            log.error("Unhandled message type received:  %r %r" % (
+                msg['topic'], msg['body']['msg_id']))
+
+    def _consume_full_build(self, msg):
+        """ This is called when we receive a message matching the newer pungi4
+        full build topic.
+        """
+
+        builds = list()  # These will be the Koji task IDs to upload, if any.
+
+        msg = msg['body']['msg']
+        if msg['owner'] != 'releng':
+            log.debug("Dropping message.  Owned by %r" % msg['owner'])
+            return
+
+        if msg['instance'] != 'primary':
+            log.info("Dropping message.  From %r instance." % msg['instance'])
+            return
+
+        # Don't upload *any* images if one of them fails.
+        if msg['new'] != 1:
+            log.info("Dropping message.  State is %r" % msg['new'])
+            return
+
+        # Create a Koji connection to the Fedora Koji instance to query.
+        koji_session = koji.ClientSession(fedimg.KOJI_SERVER)
+        children = koji_session.getTaskChildren(msg['task_id'])
+        for child in children:
+            if child["method"] == "createImage":
+                builds.append(child["id"])
 
-        log.info('Received %r %r' % (msg['topic'], msg['body']['msg_id']))
+        if len(builds) > 0:
+            fedimg.uploader.upload(self.upload_pool,
```

Comment:
Might as well, yes.  There are some log statements inside service.upload... but they get called later on.  Could be useful to log the urls here too, yes.  I'll do it as a separate PR.